### ISSUE TITLE
TWRP can't mount /system rw after installing GSI (Samsung galaxy s9)

### ISSUE THREAD

**Issue by ciucciodifuoco** [phhusson/treble_experimentations#1617]
Pretty much as the title says. 

After flashing this GSI image (ab version, tried both VNDKLite, vanilla and gapps) TWRP stops being able to mount the /system partition in rw mode. This makes impossible to flash any zip that needs the system partition. If i restore the previous rom (stock rom) backup, twrp regain the features and everything works normally. 

Just for clarification: i already disabled dm-verity and encryption by flashing the appropriate zip before installing the gsi image. In fact the rom boots normally and works fine

The error twrp 3.4 shows is :

Can't mount /system_root (structure needs cleaning)

Also tried twrp 3.1.1, but the error remains the same
Any help or hint to fix this issue?

**Comment by phhusson** [comment]
you're really sure it doesn't work with vndk lite variant?

Le mer. 9 déc. 2020 à 22:50, ciucciodifuoco <notifications@github.com> a
écrit :

**Comment by ciucciodifuoco** [comment]
Yes, it's the last variant i tried. To be 100% sure i will try again after wiping everything, but i can already answer YES to your question. Will gt back to you in 10 mins

**Comment by ciucciodifuoco** [comment]
I confirm again the previous comment: i flashed everything again after wiping all what could be wiped, the rom boots fine, but when i get again in twrp and go in the Mount section, pressing on the check near System does nothing. If i also check "mount system partition read-only" then it works ok. 

i'll be willing to test any proposed fix :)

**Comment by phhusson** [comment]
k well get me recovery's kernel logs after trying to mount it rw

Le mer. 9 déc. 2020 à 23:04, ciucciodifuoco <notifications@github.com> a
écrit :

**Comment by ciucciodifuoco** [comment]
Find both dmesg and kernel logs attached, i pressed the check to mount the system exactly 3 times:
[dmesg.log](https://github.com/phhusson/treble_experimentations/files/5669044/dmesg.log)
[recovery.log](https://github.com/phhusson/treble_experimentations/files/5669045/recovery.log)

**Comment by phhusson** [comment]
```
<3>[    8.812397]  [5:       recovery: 3507] EXT4-fs (sda18):
ext4_check_descriptors: Block bitmap for group 0 overlaps block group
descriptors
<3>[    8.812409]  [5:       recovery: 3507] EXT4-fs (sda18): group
descriptors corrupted!
```
k great, looks like there is a bug either in e2fsprogs or your ext4 kernel
driver...
Well I'll try to think of something, though it'll take time.

Le mer. 9 déc. 2020 à 23:12, ciucciodifuoco <notifications@github.com> a
écrit :

**Comment by ciucciodifuoco** [comment]
As a matter of fact, i just tried with the Google official GSI image (aosp_arm64-exp-RP1A.200720.009) and there is the same bug. I was coming from a custom rom based on the stock samsung one ui , don't know if it can make any difference.

BTW thanks for the help and as i said i'll be willing to test any possible fix

**Comment by ciucciodifuoco** [comment]
Also take a look at this:
https://www.suse.com/support/kb/doc/?id=000019314

**Comment by ciucciodifuoco** [comment]
This should be the fix you are looking for:

https://review.lineageos.org/c/LineageOS/android_kernel_motorola_msm8974/+/274962

**Comment by phhusson** [comment]
I can't really ask users to update their kernel ;-)
So I'll make a better vndk lite variant, that isn't weird like that

Le mer. 9 déc. 2020 à 23:43, ciucciodifuoco <notifications@github.com> a
écrit :

**Comment by Snagap** [comment]
Hi, This is a bit late but I can advise I have found that twrp 3.3.1-0 allowed the caos gsi system partition to be mounted RW. I expect it would allow the phh gsi as well as caos is based on phh. If that is all you need then maybe that is worth trying?

I also tried the google arm64 ab gsi as it needs the system partition resized to allow gapps to be installed. But it still only allowed mount as read only. However I did find a weird work around that solved this issue. Starting with twrp 3.5.2_9-0 I wiped cache, dalvik, system and data, rebooted into twrp and installed the gsi. Then booted the gsi. Then shutdown and booted back into twrp. I then installed twrp 3.3.1-0 and rebooted into twrp. Selecting the system partition as mount RO I then made a back-up of the system partition. Then I re-installed twrp 3.5.2_9-0 and rebooted into twrp. Then I wiped the system partition and rebooted into twrp, then I restored the system partition from the back-up and rebooted the ROM. It booted fine. Then when I rebooted back into twrp the system partition could be mounted RW and the system partition size was full allowing me to install gapps. I found that trying to do the restore of the backed up system partition whilst still in twrp 3.3.1-0 resulted in a boot loop when I booted up the restored ROM and I got similar results with using twrp 3.4.0-0 and 3.5.0_9-0. This may be because after 3.5.0_9-0 the ability to back-up the system partition is not in the S9 twrp anymore, only the system image, so it has been built differently which for some reason enables this process? 

I then used this procedure on the caos gsi and it worked fine for me. I then installed Bitgapps and restored my data and the phone so far seems to be okay. I can now mount system as RW and the error saying system_root can't be mounted, file system needs cleaning is gone when using 3.5.2_9-0 and I can resize and repair the system partition with no errors showing.

Whilst this process worked for me I can't guarantee this will work for anyone else. Maybe I got lucky with things this time round, so you do so at your own risk. If you do anything make sure you have a back-up of your data first. 

Apologies for the lengthy post but thought the detail may be of interest.

**Comment by Rocky7842** [comment]
For someone is facing the issue, this may (or may not) help:
1. Tick the "mount system partition read-only" box in mount menu and mount system.
2. Backup "system" (not "system image").
3. Restore the backed-up "system".
4. Now you can mount it r/w.

Note: On most phones you can backup and restore "system" without issue. But afaik on some Samsung phones it may be unbootable if you restore "system" rather than "system image".
